devel/style_guide.rst

```diff
@@ -0,0 +1,144 @@
+Zeek Coding Style and Development Guide
+=======================================
+
+Historical Context
+------------------
+
+The Zeek codebase is quite old as code goes, with the original code being written in 1995. This means that a large portion of it was written before a lot of modern C++ existed. As such, a lot of the early design decisions were made in that context. This guide strives to suggest modern techniques where possible, even when the existing code doesn't follow it to the letter. All new code should follow this guide. Old code should be updated to follow this guide when it is modified. 
+
+Coding conventions and style guide
+----------------------------------
+
+Basic Formatting and Indentation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The base formatting for all Zeek C++ code, including C++ code inside BIF files, follows the Whitesmith coding style (https://en.wikipedia.org/wiki/Indentation_style#Whitesmiths_style).
+
+Tabs vs. Spaces
+^^^^^^^^^^^^^^^
+
+Use tabs for indentation and spaces for alignment.
+
+File naming
+^^^^^^^^^^^
+
+Header files should always use the `.h` extension. Implementation files should always use the `.cc` extension.
+
+Include guards
+^^^^^^^^^^^^^^
+
+All headers should have an `#infdef`/`#define` include guard at the start of it to avoid duplicate includes. Do not use `#pragma once`.
```

Comment:
I think include guards vs #pragma comes down to a preference thing. I generally prefer include guards since they help with things like symlinks (since #pragma can't tell that a file is the same), but given that we're not doing that sort of thing in Zeek maybe #pragma is ok?

There's always the argument that #pragma isn't a standard, but any compiler we're ever going to use is going to implement it.

Reply:
I've come to like `#pragma once` quite a bit. It looks horrible and is non-standard, but it's so much easier than include guards. We should be consistent, but I'd be fine moving over to it.

Reply:
I'd be fine with that as well.

Reply:
Despite being non-standard, it is support by all compilers relevant to the Zeek community. From our experience, the advantages in robustness, productivity, and maintainability over include guards trump the standardization aspect.